File: 021. Merge Two Sorted Lists/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def mergeTwoLists(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        if l1 == None:
            return l2
        if l2 == None:
            return l1

        lists = None

        if l1.val <= l2.val:
            lists = ListNode(l1.val)
            l1 = l1.next
        else:
            lists = ListNode(l2.val)
            l2 = l2.next
        Head = lists

        while (l1 != None and l2 != None):
            if l1.val <= l2.val:
                lists.next = ListNode(l1.val)
                lists = lists.next
                l1 = l1.next
            else:
                lists.next = ListNode(l2.val)
                lists = lists.next
                l2 = l2.next

        while (l1 != None):
            lists.next = ListNode(l1.val)
            lists = lists.next
            l1 = l1.next
        while (l2 != None):
            lists.next = ListNode(l2.val)
            lists = lists.next
            l2 = l2.next

        return Head

File: 021. Merge Two Sorted Lists/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
{
  public:
    ListNode *mergeTwoLists(ListNode *l1, ListNode *l2)
    {
        if (l1 == NULL)
        {
            return l2;
        }
        if (l2 == NULL)
        {
            return l1;
        }

        ListNode *list = NULL;

        if (l1->val <= l2->val)
        {
            list = new ListNode(l1->val);
            l1 = l1->next;
        }
        else
        {
            list = new ListNode(l2->val);
            l2 = l2->next;
        }

        ListNode *head = list;

        while (l1 != NULL && l2 != NULL)
        {
            if (l1->val <= l2->val)
            {
                list->next = new ListNode(l1->val);
                l1 = l1->next;
                list = list->next;
            }
            else
            {
                list->next = new ListNode(l2->val);
                l2 = l2->next;
                list = list->next;
            }
        }

        while (l1 != NULL)
        {
            list->next = new ListNode(l1->val);
            l1 = l1->next;
            list = list->next;
        }

        while (l2 != NULL)
        {
            list->next = new ListNode(l2->val);
            l2 = l2->next;
            list = list->next;
        }
        return head;
    }
};